src/cargo/core/compiler/build_runner/compilation_files.rs

```diff
@@ -604,12 +604,18 @@ fn compute_metadata(
 
     METADATA_VERSION.hash(&mut shared_hasher);
 
-    // Unique metadata per (name, source, version) triple. This'll allow us
-    // to pull crates from anywhere without worrying about conflicts.
-    unit.pkg
-        .package_id()
-        .stable_hash(bcx.ws.root())
-        .hash(&mut shared_hasher);
+    if unit.is_std {
+        // SourceId for build-std crates is a full absolute path inside the sysroot.
+        // Don't hash it to avoid the metadata hash changing depending on where the user installed rustc.
+        unit.pkg.name().hash(&mut shared_hasher);
+    } else {
+        // Unique metadata per (name, source, version) triple. This'll allow us
+        // to pull crates from anywhere without worrying about conflicts.
+        unit.pkg
+            .package_id()
+            .stable_hash(bcx.ws.root())
```

Comment:
Another approach to this: if a unit is from `-Zbuild-std`, pass its workspace root path to `stable_hash` instead.

Reply:
that's cleaner yes! I've changed the code to do this.

Reply:
Sorry but I got paranoid that this could be applied to the same logic, for consistency.

https://github.com/rust-lang/cargo/blob/99dff6d77db779716dda9ca3b29c26addd02c1be/src/cargo/core/compiler/build_runner/compilation_files.rs#L203-L206

(It doesn't really matter though. The hash is for target directory layout for fingerprint and build-script output…)

Reply:
Never mind. It needs more tweaks and seems meaningless.